### DISCUSSION TITLE
Selecting a sentence, part of which is highlighted or annotated

### DISCUSSION THREAD

**Discussion by dummifiedme** [pymupdf/PyMuPDF#828] [Q&A] [answered]
Wanted to discuss,  if there is a way we can have sentences extracted from a pdf, part of which is highlighted or annotated.

It might be off topic, but would like to know if someone has ideas on how to get thats sort of a thing done. 
I tried exploring 'nltk' library. But integrating it with the PyMuPDF way of things, seems to be difficult?

If someone has done some work in that direction, please share :)

**Comment by JorjMcKie** [comment]
Highlighting takes places in the form of annotations.
And how text can be extracted, that is contained in annotation rectangles has already been discussed elsewhere. It basically works by extracting text from a series of rectangles defined by the annotation rectangles.
Of course there is no way to know if the thus extracted text makes up a complete sentence - if that was the direction of your question.
Suchthings are out of PyMuPDF scope and more a matter of semantics or some AI.

The opposite, namely how to locate and then highlight text that spans across several lines, is being discussed in some detail in [here](https://github.com/pymupdf/PyMuPDF-Utilities/tree/master/word%26line-marking).

**Reply by dummifiedme** [reply]
Okay. Got it. 
I was thinking on the lines of appraching it like:
- look for highlighted text. 
- Select the first word from the highlights
- Check for the previous word in the wordlist 
- check till t is a `.` or `word.`
- Join them all and similiarly for the other end.

For simple use case, can something like this possible or I am just ignoring all the things there is?

And yes, I know this is beyond the scope PyMuPDF :) I hope you don't mind.

Regarding the opposite, yes, I had gone through that link. :D

**Reply by JorjMcKie** [reply]
> I hope you don't mind.

Of course not.
And yes: backward and forward searching until start / end of sentence is found, may work / can word / should work.
But only if a number of conditions are true or, vice versa, some complications can be excluded:
* Page layout only has one text column
* extracted text pieces of the highlights can neatly be matched with extracted text, via getText("words"), or the spans of getText("dict")
* partly overlapping the previous: what happens, if text fonts / fontsizes change within a line, or text pieces only **_roughly_** have the same baseline, but actually do differ by 1 or 2 pixels, ... In such cases some cleaning / simplification has to happen first so that these portions are forced on the same line and in the proper sequence

**Reply by dummifiedme** [reply]
Yes, there would be complications, I agree. 
I thought the word list which `getText(“words”)` captures is like a chain of words. It doesn’t really have any format right? It’s essentially just a list of words? So, if it is so, from what I could understand, I can traverse through it word by word and find the ends of sentences. 
But yes, as you said, there might be complications due to changing fonts or inability of the function to capture the words properly. Again, that inability will anyway effect if I am capturing the highlight. I mean, that “inability” would be at both the places, in that case we can’t rely on extractions any way (ie in case of a “bad” PDF).

**_What is the difference between different arguments being passed in `getText()`? I only used “words” till now. What does “dict” do? And what are the other arguments that could be used?_** 
I will try to read over in the documentation in my weekends maybe 👍 

_**Also, does `getText(“words”)` capture a line of words without considering the columns?**_ I tried it with a PDF with boxed contents, there it kind of stopped at the edge of the box and started to capture next line. It captured the boxed text later on. I thought it was working on something like “Blocks”, I don’t know if its really the case.

**Reply by JorjMcKie** [reply]
> Also, does getText(“words”) capture a line of words without considering the columns?

Yes, exactly. But each word item also contains block number, line number in the block, word number in the line. So you would presumably sort the words according to their page position (using the 4 leading bbox floats). If things are not too awkward, then pages with multiple text columns would be expected to have different block numbers for different columns. So block number would be a good candidate as the leading sort key. Of course, a text column may still consist of more than one block.

**Reply by dummifiedme** [reply]
Got it. I still have to learn about `bbox`. But I get what you are saying. To have a more robust and fail safe code, I should have a sound sorting. Currently, I just loop around the annotations and capture whatever comes up. So far so good on my documents. AS you had suggested somewhere in the discussion, I will have to eventually sort the things according to their positions. I will at some point :)

**Comment by JorjMcKie** [comment, marked as answer]
I know that @zeusrb is doing stuff in that direction. Maybe he wants to jump in and share his position or findings.

**Reply by zeusrb** [reply]
Thanks Jorj for pointing to me I still owe you a wiki!
What I did is highlight a pdf working with rectangle containing a sentence.

At first stage I was parsing tokens with the page.search() until I found that it was failing on words with accents (in italian used very often).
At second stage, I've made in this way (sorry for the pseudo code).

1. simplify the text string you are searching (tokenize, clean stopwords) in strings made of about 200 words (number of words is up to you). final chunks are made of less words.
2. keep track of what you are doing in a pd.Dataframe with ID, original chunk, tokenized chunk
3. read the pdf, pretty formatted PDF, meaning, not to be OCRed or with fancy object inside).
 Important is when you get text (follow my conversation with Jorj and you will find the piece of code): 
 _flags = fitz.TEXT_PRESERVE_LIGATURES | fitz.TEXT_PRESERVE_WHITESPACE_
 and
 when you return the text:
 _line_list, text_pg.encode(ENCODING, 'replace').decode(ENCODING)_
4. create the same dataframe schema with sentences coming from the read file plus the page number the chunk is coming from.
5. 5. now make the cartesian product of the two dataframes (dangerus operation for big pdf files ....)
6. compare the chunks with the method you like (difflib.SequenceMatcher, cosine similarity, nltk, ...)
7. while you iterate in the (maybe huge) dataframe keep track of max scores made by the sentence you ar searching and the piece of text coming from the page.
8. choose a threshold and filter the dataframe, at that point you will have a short list of chunk of string to be searched combined with string of the document with positions in your document firt of al the page then remember fitz can give you all the rectangles positions when you read it first time.
9. Whit rows of the dataframe with highest score (above your choose threshold), I start searching 3 (up to you) words at time inside the page chunk matched with the document text rows storing Upper Left rectangle starting point A of first good line and the LR point B of last found.
10. addHighlight annotation between point A to point B (and because I need addTextAnnotation aside of PointA with some info)  

Not perfect not super fast (because my dirty code not because pyMupdf.... ;-) ), but working, the key point is the matching.... and that I'll start soon applying some AI stuff
![image](https://user-images.githubusercontent.com/29802059/104379095-4388f000-5529-11eb-862c-857e4327d338.png)

**Reply by zeusrb** [reply]
BTW, I' keeping the good chunks to retrofit my AI engine, so next round will do better  search and document categorization ... ;-)

**Reply by dummifiedme** [reply]
It will take some days if not more to wrap my head around it. I hope I can take this up long term as a hobby along with my studies.

Thanks to PyMuPDF, my workflow became a bit smooth.

The ability to get the sentences from incomplete highlights would be something I would like to have sometime in future.
PS It's really interesting to learn, but the Exams!😐

**Reply by JorjMcKie** [reply]
What subjects are you studying?

**Reply by dummifiedme** [reply]
I am preparing for administrative services. So it involves basically everything :)
If I have to list the subjects, it would be somewhat around: 
- History (Ancient India, Medieval India, Modern India, After Independence, World History)
- Political Science and governance (Mostly about Indian Polity)
- Economics (again revolves around Indian economy and other economic policies that might in turn affect India, eg oil)
- Geography and Disaster Management ( physical, human and Indian geography)
- International Relations (Mostly around indian relations and policies that effect India)
- Art and Culture (Architecture, Dances,, etc)
- Environment Studies
- Defence and Security
- Ethics and the related aptitude.
- And of course Science :p 

Other than these, we need to be aware of current developments. And an optional subject which is completely personal choice. I chose Anthropology (despite being a mechanical design engineer 😉 )

So yeah, that’s about it.

**Reply by JorjMcKie** [reply]
Uff, not really can be called a sort list 🙄

**Comment by navjeetc** [comment]
I have a similar requirement to extract only highlighted text from multiple pages. It is not clear to me from above if that can be successfully done. If yes, a small snippet of code would be nice.

**Reply by JorjMcKie** [reply]
Depending on how "highlighting" has been done technically, this can be easy or complex.
The easy case:

* Iterate over the annotations of a page and down-select the highlight annotations.
* Take the annotation rectangle and extract the text covered by it.

```python
for annot in page.annots(types=[fitz.PDF_ANNOT_HIGHLIGHT]):
    rect = annot.rect  # annotation rectangle
    rect += (-5, -5, 5, 5)  # enlarge it a bit to cope with imprecisions
    print(page.get_textbox(rect))  # the text
```

If this does not deliver as expected, the highlighting may have been done by vector graphics.
Please ping me if you need more support.

**Reply by navjeetc** [reply]
This is very helpful. Thanks.

**Reply by JorjMcKie** [reply]
my invitation stands...
